iroh-net/src/magic_endpoint.rs

```diff
@@ -560,10 +560,40 @@ impl MagicEndpoint {
     ///
     /// Returns an error if closing the magic socket failed.
     /// TODO: Document error cases.
-    pub async fn close(&self, error_code: VarInt, reason: &[u8]) -> Result<()> {
-        self.cancel_token.cancel();
-        self.endpoint.close(error_code, reason);
-        self.msock.close().await?;
+    pub async fn close(self, error_code: VarInt, reason: &[u8]) -> Result<()> {
+        let MagicEndpoint {
+            msock,
+            endpoint,
+            cancel_token,
+            ..
+        } = self;
+        cancel_token.cancel();
+        endpoint.close(error_code, reason);
+
+        {
+            // waiting for connections to close can be slow: this timeout decide whethers we should
+            // inform the user of a slow operation
+            let logging_timeout = tokio::time::sleep(Duration::from_millis(500));
+            futures::pin_mut!(logging_timeout);
+
+            let wait_idle = endpoint.wait_idle();
+            futures::pin_mut!(wait_idle);
+
+            // race the timeout and closing. Nothing else needs to be done if closing finishes
+            // first. If the timeouts is done first, give the user some logs
+            if let future::Either::Right(((), wait_idle_fut)) =
+                future::select(wait_idle, logging_timeout).await
```

Comment:
Given we want to get rid of future sometime maybe using `tokio::select!` is better?  It's also more often used in our code.  Feel free to disagree however.

Reply:
I had always used `tokio::select!` before, until Rudiger for some things in another repo asked me to use the `futures` function. For such small cases I know prefer it for a couple of reasons:
1. no macro, so the code inside is correctly formatted by `cargo fmt`. Of course this is a lesser reason
2. Saves me dealing with cancellation safety. `future::select` returns the future that didnt' complete, which is exactly what I'm taking advantage of here. `tokio::select!` would simply drop the future, we would likely need to have this in a loop and carefully think how to wrap the future so that it's not just recreated on the iteration - because I doubt `wait_idle` is cancel-safe -

Of course we don't need to deal with any of this if we just log it all the time

Reply:
removed in c1238e3, going with `debug` as a compromise